app/models/user.rb

```diff
@@ -1,11 +1,49 @@
 class User < ApplicationRecord
-    before_save { self.email = email.downcase }
-    VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
-    validates :name,  presence: true, length: { maximum: 50 }
-    validates :email, presence: true, length: { maximum: 255 },
-                      format: { with: VALID_EMAIL_REGEX },
-                      uniqueness: { case_sensitive: false }
-
-    has_secure_password
-    validates :password, presence: true, length: { minimum: 6 }
+  attr_accessor :activation_token
+  before_create :create_activation_digest 
+  before_save :downcase_email
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :name,  presence: true, length: { maximum: 50 }
+  validates :email, presence: true, length: { maximum: 255 },
+                    format: { with: VALID_EMAIL_REGEX },
+                    uniqueness: { case_sensitive: false }
+
+  has_secure_password
+  validates :password, presence: true, length: { minimum: 6 }, allow_nil: true
+
+  def self.digest(string)
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+                                                  BCrypt::Engine.cost
+    BCrypt::Password.create(string, cost: cost)
+  end
+
+  def self.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  # Will returns true if the given a token matching the digest.
+  def authenticated?(attribute, token)
+    digest = send("#{attribute}_digest")
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password?(token)
+  end
+
+  def activate
+    update_columns(activated: true, activated_at: Time.zone.now)
+  end
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
```

Comment:
always when working with mailers, user deliver_later

```suggestion
    UserMailer.account_activation(self).deliver_later
```

this queues the email sending and when resources are available, the mail is sent asynchronously.

Reply:
ok. Thanks###